text/0000-arbitrary_enum_discriminant.md

```diff
@@ -0,0 +1,268 @@
+            }
+            // This shouldn't be needed.
+            AnimationValue::InlineSize(ref void) => {
+                match *void {}
+            }
+            AnimationValue::TransformOrigin(ref origin) => {
+                do_something_with_transform_origin(origin)
+            }
+        }
+    }
+}
+```
+
+We suggest generalising the explicit discriminant notation to all enums,
+regardless of whether their variants have fields or not:
+
+```rust
+#[repr(u16)]
+enum AnimationValue {
+    Color(Color) = LonghandId::Color as u16,
+    Height(Length) = LonghandId::Height as u16,
+    TransformOrigin(TransformOrigin) = LonghandId::TransformOrigin as u16,
+}
+
+impl AnimationValue {
+    fn id(&self) -> LonghandId 
+        // We can use the unsafe trick again.
+        unsafe { *(self as *const Self as *const LonghandId) }
+    }
+
+    fn do_something(&self) {
+        // No spurious variant anymore.
+        match *self {
+            AnimationValue::Color(ref color) => {
+                do_something_with_color(color)
+            }
+            AnimationValue::Height(ref height) => {
+                do_something_with_height(height)
+            }
+            AnimationValue::TransformOrigin(ref origin) => {
+                do_something_with_transform_origin(origin)
+            }
+        }
+    }
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An enumeration with only field-less variants can currently have explicit
+discriminant values:
+
+```rust
+enum ForceFromage {
+    Emmental = 0,
+    Camembert = 1,
+    Roquefort = 2,
+}
+```
+
+With this RFC, users are allowed to put explicit discriminant values on any
+variant of any enumeration, not just the ones where all variants are field-less:
+
+```rust
+enum ParisianSandwichIngredient {
+    Bread(BreadKind) = 0,
+    Ham(HamKind) = 1,
+    Butter(ButterKind) = 2,
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+The production for enumeration items becomes:
+
+```
+EnumItem :
+    OuterAttribute*
+    IDENTIFIER ( EnumItemTuple | EnumItemStruct)? EnumItemDiscriminant?
+```
+
+## Semantics
+
+The limitation that only field-less enumerations can have explicit discriminant
+values is lifted, and no other change is made to their semantics:
+
+ * enumerations with fields still can't be casted to numeric types
+   with the `as` operator;
+ * if the first variant doesn't have an explicit discriminant,
+   it is set to zero;
+ * any unspecified discriminant is set to one higher than the one from
+   the previous variant;
+ * under the default representation, the specified discriminants are
+   interpreted as `isize`;
+ * two variants cannot share the same discriminant.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This introduces one more knob to the representation of enumerations.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
```

Comment:
I think you have encountered the XY Problem. You are finding a solution to the problem that different enums don't give guarantees about their variant indices, when what you originally wanted was converting enums whose variants have fields into enums whose variants have no fields.

So here's my alternative:

Allow annotating enums with `#[discriminant(OtherEnum)]`, where `OtherEnum` must have a variant for each variant in the annotated enum (with matching name). The annotated enum will then take all its discriminant values from `OtherEnum`.

`as` casts could then convert from annotated enums to their discriminant enum.

Reply:
That seems more convoluted, IMO. Definitely harder to implement, at least.

Reply:
Sure, but it has a path forward for having this operation in safe code.

Alternatively a procedural macro could generate said conversions with this RFC. So the compiler magic might not be necessary.

Reply:
This would be great and nice, if only real world allowed me to do that.

I simplified my use case for the sake of the RFC: `PropertyDeclaration` actually has *more* variants than `LonghandId`, thanks to custom CSS properties (and some other stuff that I don't need to mention here):

```rust
#[repr(u16)]
enum PropertyDeclaration {
    Color(Color),
    Height(Length),
    InlineSize(Length),
    TransformOrigin(TransformOrigin),
    Custom(CustomDeclaration),
}

struct CustomDeclaration {
    name: Name,
    value: Value,
}

pub enum PropertyDeclarationId<'a> {
    Longhand(LonghandId),
    Custom(&'a Name),
}

impl PropertyDeclaration {
    #[inline]
    pub fn id(&self) -> PropertyDeclarationId {
        if let PropertyDeclaration::Custom(ref declaration) =  *self {
            return PropertyDeclarationId::Custom(&declaration.name);
        }
        let id = unsafe { *(self as *const _ as *const LonghandId) };
        PropertyDeclarationId::Longhand(id)
    }
}
```

In general, I think this alternative is too specific to the exact use case described in the RFC, and cannot fulfil more intricate ones like the actual stuff I require in Servo, or use cases I could imagine for this feature with FFI, @Gankro may have an opinion on that regard here.

Reply:
You mention in the RFC that rust is generating a 4KB jump table for the required match expression - this seems excessive, even if the discriminants don't match exactly. How many variants does this enum have? Maybe part of the solution should be improving the code that rust generates in this case.

Reply:
@Diggsey There are [>320](https://doc.servo.org/geckolib/style/properties/enum.PropertyDeclaration.html) variants, so far. I welcome any rustc improvement but they won't be able to *remove* all jump tables if the discriminants of the common variants to `AnimationValue` and `PropertyDeclaration` don't coincide.

Reply:
@Diggsey Also if you are curious, [here](https://github.com/servo/servo/pull/19956) is the PR where `PropertyDeclaration::id` was slimmed down from 4KB to mere 96 bytes, and [there](https://github.com/servo/servo/pull/20014) is the PR where I simplified `AnimationValue::id`.